### ISSUE TITLE
Loss of content tree image type for container downloaded but not loaded during reboot

### ISSUE THREAD

**Issue by deitch** [lf-edge/eve#3704]
As discussed with @eriknordmark and @OhmSpectator 

## Description of the error

If a container image (not a disk VM) was downloaded and in verification or verified, but not yet loaded into the content-addressable-store, i.e. containerd, and the device reboots, upon reboot it loads the image incorrectly into containerd, thinking it is a disk. It then tries to create a snapshot and fails with an error like:

```
failed to extract layer sha256:3a46589e93913901d4487d6964fa82deeb7cfbed293f3be00ca27da8556f6aa9: failed to get stream processor for application/vnd.lfedge.disk.layer.v1+raw: no processor for media-type: unknown"
```

The key indicator in the logs is the above, when we know that this digest is for a a container image, not a disk, yet it believes it to be a disk by media-type. It then tries to build a snapshot with it, which cannot be done except with a layer.

## Cause

1. The device reboots
2. Tries to restore lots of stuff, especially out of `/persist`. Mostly, this works fine.
3. Sees that there is a content tree in download process request (or does anew, it doesn’t matter)
4. Tries to find `BlobStatus` for the root blob using `lookupOrCreateBlobStatus()` [here](https://github.com/lf-edge/eve/blob/3db87c6f9806a5ae5e565c21299f638694523a44/pkg/pillar/cmd/volumemgr/handlecontent.go#L192). Function is defined [here](https://github.com/lf-edge/eve/blob/master/pkg/pillar/cmd/volumemgr/blob.go#L403).
5. It looks up the blob status [here](https://github.com/lf-edge/eve/blob/master/pkg/pillar/cmd/volumemgr/blob.go#L409), does not find it, so it looks for a `VerifyImageStatus` [here](https://github.com/lf-edge/eve/blob/master/pkg/pillar/cmd/volumemgr/blob.go#L419).
6. Most of the time, it does not find one, so it goes backwards, eventually redownloading; those cases are fine.
7. In the circumstance outlined above, it _does_ find a `VerifyImageStatus`, so it creates a `BlobStatus` and publishes it [here](https://github.com/lf-edge/eve/blob/master/pkg/pillar/cmd/volumemgr/blob.go#L419). This `BlobStatus` does not contain a media-type, which is defined as a property of `BlobStatus` [here](https://github.com/lf-edge/eve/blob/master/pkg/pillar/types/blob.go#L36).
8. Now we have a `BlobStatus`, but the remaining flow no longer knows the media type for that image. Of course, _we_ know that it is an OCI image manifest, but EVE does not.
9. When it comes time to determine if this is a manifest or not (i.e. a VM disk, although in the future, could be something else), and so if it needs to be wrapped with a manifest, it calls `IsManifest()` [here](https://github.com/lf-edge/eve/blob/3db87c6f9806a5ae5e565c21299f638694523a44/pkg/pillar/cmd/volumemgr/updatestatus.go#L315), defined [here](https://github.com/lf-edge/eve/blob/master/pkg/pillar/types/blob.go#L68), which checks the media-type, finds none, and says, “it definitely is not a manifest”
10. It gets wrapped by a new manifest saying it is a disk, and the rest is history

Normally, we get the media type as part of the download, it gets passed to the verify process, which then sets it on the `BlobStatus`, see [here](https://github.com/lf-edge/eve/blob/3db87c6f9806a5ae5e565c21299f638694523a44/pkg/pillar/cmd/volumemgr/blob.go#L196). It does not preserve that into the `VerifyStatus`.

However, when we restore after a reboot when the download status no longer exists, that bit of info is lost, and we recreate the `BlobStatus` _without_ adding the media-type.

## Potential solution

1. Carry the media type from the download into the `VerifyStatus`, when we set it on `BlobStatus`, which is lost on reboot. It cannot hurt.
2. On reboot, when we set the `BlobStatus` from the `VerifyStatus`, populate it from there.

The above assumes that the `VerifyStatus` persists through reboot. That may not be correct.

## Summary

One critical piece of information - a blob's media-type - is passed from the downloader (retrieves it via the content header) to the `BlobStatus`, but is lost on reboot. This works if we download all over again, or completed download+verify+load - all or none. If we pick it up again in the middle, we lose that information. We need to find a way to persist that information across reboots.

**Comment by deitch** [comment]
@eriknordmark wrote:

> Adding `ContentType` in `VerifyImageConfig.Status` seems reasonable, but can we determine this on boot when the verifier creates the `VerifyImageStatus` based on what it finds in the verified directory? (Or does such code no longer exist?)

I didn't think about that. I had assumed that since the code failed to find a `BlobStatus`, but found a `VerifyImageStatus`, that the `BlobStatus` did not persist, but the `VerifyStatus` _did_ persist.

I think you are saying that is incorrect, that `VerifyStatus` is created from the contents of the `/verifier/` subdirectories.

> can we determine this on boot when the verifier creates the `VerifyImageStatus` based on what it finds in the verified directory? 

We can guess, somewhat reliably, but it is brittle and asking for future problems.

Anything that actually is a manifest or index, if we examine it, we will determine that quickly, and find its media-type inside it. Arguably, any of these should have the media-type, but in the [spec for manifest](https://github.com/opencontainers/image-spec/blob/main/manifest.md#image-manifest-property-descriptions) and [spec for index[(https://github.com/opencontainers/image-spec/blob/main/image-index.md) it is a "SHOULD", not a "MUST".

We also could do things like try to parse it and see if it works, but that is brittle and fragile.

Further, if the piece that is verified but not loaded is just part of something else (image layer), it is the part _surrounding_ it (i.e. manifest) that knows what media-type it is.

Essentially, the rule of thumb is: the _pointer_ at something tells you its media-type, rather than the thing itself.

If you are saying that the _only_ thing that persists is the verified or verifying blob, and not the status, then the solution is incomplete. We would need to update verifier to store the content type along with the blob itself in `/verifier/`

> It seem code elsewhere where we look for a `DownloaderStatus` (and use `setBlobTypeFromContentType`) and the look for a `VerifyImageStatus` (and do not set the Type). Should we put the setting of `MediaType` in `updateBlobFromVerifyImageStatus` to make it catch both/all?

I don't understand.

The cases I do understand for post-restart are:

* none case (no download) - get content type, pass it all, it works
* all case (completely loaded into containerd) - already there, it works
* download complete, not verified or verified but not loaded - have the blob, not the content-type, needs to be fixed at both the "create BlobStatus" level, as in my proposal above, but also in storing the content type in persist, as you pointed out
* download incomplete, or not yet moved to `/verifier` dir - I think this is what you are referring to above, but I am unsure.

**Comment by eriknordmark** [comment]
> I think you are saying that is incorrect, that `VerifyStatus` is created from the contents of the `/verifier/` subdirectories.


Corrrect. verifyImageStatusFromImageFile() is using the fields in the filename (plus the size from Stat()) to create the status. We could change the scheme for the filename to include the content type. If we do that we might want to handle the currrent filename scheme in case there are old files on EVE update in the verified directory.

**Comment by deitch** [comment]
Then we probably should fix this in 2 stages. First, update the verifier (and downloader?) to include media-type information with the file. One of:

* Changing the name to `abcdef11212.media-type` (although media-types usually contain `/` and sometimes other characters in it, so it would prob need to be base64-encoded)
* Adding a file next to it, like `abcdef112212.media-type` where the _contents_ of this file are the media-type?

As you said, anything that parses it would have to properly handle it and be backwards-compatible.

Second, once that is in place, we can update whatever reads it to read the media-type as well, and add it to `VerifyStatus` and `BlobStatus`.

**Comment by deitch** [comment]
Actually, the order doesn't matter, as long as either can handle both with and without the media-type.

Thoughts on where it should go? Filename extension or sidecar file?

**Comment by deitch** [comment]
I don't think that would be enough. We need to capture the actual media type.

**Comment by eriknordmark** [comment]
Do we need the media-type/content-type string or is it sufficient to encode the two? values we care about as integers? I haven't looked at how we use the value from the blobStatus but I recall seeing a case statement with just two cases.

Having it as part of the filename is simpler than a separate file in that we don't need to track down where we need to delete that separate file, and no need to handle the case of the separate file being empty (due to ext4 logging of meta-data but not data + a power outage shortly after file was written.)

**Comment by deitch** [comment]
I think it is just one, not two. I would use the actual information. While it is somewhat reasonable to assume eve will be updated with new content type, we cannot guarantee that. It makes ore sense for eve to remain flexible and just pass it on as is.

OK, so let's go for it in the filename. I can try and PR it, see how much I remember about this part.

**Comment by deitch** [comment]
> If we are going this path, we lose the following fields in the blob status:

@OhmSpectator how would we lose more than we do now, by adding data?

**Comment by OhmSpectator** [comment]
We are not loosing something new. We can just face some new issues if not start restoring it... Maybe it's just a case of a new race. Just in theory.

**Comment by OhmSpectator** [comment]
Now, we see a problem with the MediaType field. We have a case where EVE fails because of a lack of this info. But, as we see from the code, we are losing not only this info during the reboot. And if we fix only the MediaType field - we can face another problem later with any of the fields mentioned above. But ok, I guess we should not overcomplicate and try to resolve the problem that we have not faced yet. =)

**Comment by OhmSpectator** [comment]
In the case of this bug, I would just think about saving the rest of the fields of BlobStatus, not the entire EVE state. If it can be solved in a cheap way all along with the MediType, then why not handle it right now? But if it requires some other approach that slows down the fix of exactly this bug - I would also postpone it. 
The entire EVE state is too complex to be resolved in the scope of this bug. That's why I created the discussion that you have already seen.

**Comment by deitch** [comment]
Agreed. Now, if only it were easy to figure out where the filename is set, state is saved, etc.

**Comment by OhmSpectator** [comment]
Isn't it the place you are talking about?..
https://github.com/lf-edge/eve/blob/742676d2dd15d09de49a96fda72c5987bcee28f7/pkg/pillar/cmd/verifier/verifier.go#L595
Here, we set up the filename for VerifierStatus after verification is done, and that is the case of our interest. But still need to understand whether it's used during the reboot.

**Comment by deitch** [comment]
That is part of it. I think the call to `ImageVerifierFilenames()` [here](https://github.com/lf-edge/eve/blob/742676d2dd15d09de49a96fda72c5987bcee28f7/pkg/pillar/cmd/verifier/verifier.go#L565) might also be part of it, but there is some comment about "ensuring the filename is only sha256" once verified (see the "Blame" view for the commit). We probably can accept a change there.

I guess I can start a PR just on that part, and we can use the comments there to figure out where else it is consumed?
